File: lib/core/models/chunk.dart
import 'dart:typed_data';

import 'package:equatable/equatable.dart';

import '../constants/nfar_format.dart';
import '../utils/binary_reader.dart';
import '../utils/binary_writer.dart';

/// Represents a single chunk of archive data to be stored on an NFC tag.
class Chunk extends Equatable {
  /// Creates a new chunk.
  const Chunk({
    required this.archiveId,
    required this.totalChunks,
    required this.chunkIndex,
    required this.payload,
    required this.crc32,
    this.flags = 0,
  });

  /// UUID of the archive this chunk belongs to (16 bytes)
  final Uint8List archiveId;

  /// Total number of chunks in the archive
  final int totalChunks;

  /// Index of this chunk (0-based)
  final int chunkIndex;

  /// Payload data
  final Uint8List payload;

  /// CRC32 checksum of the payload
  final int crc32;

  /// Flags (compression, encryption)
  final int flags;

  /// Whether this chunk's data is compressed
  bool get isCompressed => NfarFlags.isCompressed(flags);

  /// Whether this chunk's data is encrypted
  bool get isEncrypted => NfarFlags.isEncrypted(flags);

  /// Archive ID as UUID string
  String get archiveIdString {
    final hex = archiveId
        .map((b) => b.toRadixString(16).padLeft(2, '0'))
        .join();
    return '${hex.substring(0, 8)}-${hex.substring(8, 12)}-'
        '${hex.substring(12, 16)}-${hex.substring(16, 20)}-'
        '${hex.substring(20, 32)}';
  }

  /// Total size of this chunk when serialized
  int get totalSize => NfarHeaderSize.total + payload.length;

  /// Serialize this chunk to bytes.
  Uint8List toBytes() {
    final writer = BinaryWriter(NfarHeaderSize.total + payload.length);

    // Magic
    writer.writeBytes(Uint8List.fromList(nfarMagic));

    // Version
    writer.writeUint8(nfarVersion);

    // Flags
    writer.writeUint8(flags);

    // Archive ID
    writer.writeBytes(archiveId);

    // Total chunks
    writer.writeUint16(totalChunks);

    // Chunk index
    writer.writeUint16(chunkIndex);

    // Payload size
    writer.writeUint16(payload.length);

    // Payload
    writer.writeBytes(payload);

    // CRC32
    writer.writeUint32(crc32);

    return writer.toBytes();
  }

  /// Parse a chunk from bytes.
  ///
  /// Throws [FormatException] if the data is invalid.
  factory Chunk.fromBytes(Uint8List data) {
    if (data.length < NfarHeaderSize.total) {
      throw FormatException(
        'Data too short: expected at least ${NfarHeaderSize.total} bytes, '
        'got ${data.length}',
      );
    }

    final reader = BinaryReader(data);

    // Validate magic
    final magic = reader.readBytes(NfarHeaderSize.magic);
    if (!validateMagic(magic)) {
      throw const FormatException('Invalid magic bytes: not an NFAR chunk');
    }

    // Version
    final version = reader.readUint8();
    if (!validateVersion(version)) {
      throw FormatException(
        'Unsupported version: $version (expected $nfarVersion)',
      );
    }

    // Flags
    final flags = reader.readUint8();

    // Archive ID
    final archiveId = reader.readBytes(NfarHeaderSize.archiveId);

    // Total chunks
    final totalChunks = reader.readUint16();

    // Chunk index
    final chunkIndex = reader.readUint16();

    // Payload size
    final payloadSize = reader.readUint16();

    // Validate remaining data
    final expectedTotal = NfarHeaderOffset.payload + payloadSize + NfarHeaderSize.crc32;
    if (data.length < expectedTotal) {
      throw FormatException(
        'Data too short for payload: expected $expectedTotal bytes, '
        'got ${data.length}',
      );
    }

    // Payload
    final payload = reader.readBytes(payloadSize);

    // CRC32
    final crc32 = reader.readUint32();

    return Chunk(
      archiveId: archiveId,
      totalChunks: totalChunks,
      chunkIndex: chunkIndex,
      payload: payload,
      crc32: crc32,
      flags: flags,
    );
  }

  /// Create a copy with updated fields.
  Chunk copyWith({
    Uint8List? archiveId,
    int? totalChunks,
    int? chunkIndex,
    Uint8List? payload,
    int? crc32,
    int? flags,
  }) {
    return Chunk(
      archiveId: archiveId ?? this.archiveId,
      totalChunks: totalChunks ?? this.totalChunks,
      chunkIndex: chunkIndex ?? this.chunkIndex,
      payload: payload ?? this.payload,
      crc32: crc32 ?? this.crc32,
      flags: flags ?? this.flags,
    );
  }

  @override
  List<Object?> get props => [
        archiveId,
        totalChunks,
        chunkIndex,
        payload,
        crc32,
        flags,
      ];

  @override
  String toString() =>
      'Chunk($chunkIndex/$totalChunks, ${payload.length} bytes, '
      'archive: ${archiveIdString.substring(0, 8)}...)';
}

File: webapp/src/chunk.ts
/** NFAR v1 chunk codec. Mirrors lib/core/constants/nfar_format.dart and lib/core/models/chunk.dart. */

export const NFAR_MAGIC = new Uint8Array([0x4e, 0x46, 0x41, 0x52]); // "NFAR"
export const NFAR_VERSION = 0x01;
export const HEADER_SIZE = 28; // fields before payload
export const CRC_SIZE = 4;
export const TOTAL_OVERHEAD = HEADER_SIZE + CRC_SIZE; // 32
export const FLAG_COMPRESSED = 0x01;
export const FLAG_ENCRYPTED = 0x02;
export const MAX_CHUNKS = 65535;
export const MAX_PAYLOAD_SIZE = 65535;

export class NfarFormatError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'NfarFormatError';
  }
}

export interface Chunk {
  archiveId: Uint8Array; // 16 bytes
  totalChunks: number;
  chunkIndex: number;
  payload: Uint8Array;
  crc32: number;
  flags: number;
}

export function encodeChunk(chunk: Chunk): Uint8Array {
  if (chunk.archiveId.length !== 16) {
    throw new NfarFormatError(`Archive ID must be 16 bytes, got ${chunk.archiveId.length}`);
  }
  if (chunk.payload.length > MAX_PAYLOAD_SIZE) {
    throw new NfarFormatError(`Payload too large: ${chunk.payload.length}`);
  }
  const out = new Uint8Array(TOTAL_OVERHEAD + chunk.payload.length);
  const view = new DataView(out.buffer);
  out.set(NFAR_MAGIC, 0);
  out[4] = NFAR_VERSION;
  out[5] = chunk.flags;
  out.set(chunk.archiveId, 6);
  view.setUint16(22, chunk.totalChunks);
  view.setUint16(24, chunk.chunkIndex);
  view.setUint16(26, chunk.payload.length);
  out.set(chunk.payload, HEADER_SIZE);
  view.setUint32(HEADER_SIZE + chunk.payload.length, chunk.crc32);
  return out;
}

export function decodeChunk(data: Uint8Array): Chunk {
  if (data.length < TOTAL_OVERHEAD) {
    throw new NfarFormatError(
      `Data too short: expected at least ${TOTAL_OVERHEAD} bytes, got ${data.length}`,
    );
  }
  for (let i = 0; i < NFAR_MAGIC.length; i++) {
    if (data[i] !== NFAR_MAGIC[i]) {
      throw new NfarFormatError('Invalid magic bytes: not an NFAR chunk');
    }
  }
  const version = data[4]!;
  if (version !== NFAR_VERSION) {
    throw new NfarFormatError(`Unsupported version: ${version} (expected ${NFAR_VERSION})`);
  }
  const view = new DataView(data.buffer, data.byteOffset, data.byteLength);
  const flags = data[5]!;
  const archiveId = data.slice(6, 22);
  const totalChunks = view.getUint16(22);
  const chunkIndex = view.getUint16(24);
  const payloadSize = view.getUint16(26);
  const expectedTotal = HEADER_SIZE + payloadSize + CRC_SIZE;
  if (data.length < expectedTotal) {
    throw new NfarFormatError(
      `Data too short for payload: expected ${expectedTotal} bytes, got ${data.length}`,
    );
  }
  const payload = data.slice(HEADER_SIZE, HEADER_SIZE + payloadSize);
  const crc = view.getUint32(HEADER_SIZE + payloadSize);
  return { archiveId, totalChunks, chunkIndex, payload, crc32: crc, flags };
}
